src/logic/player.cc

```diff
@@ -1264,116 +1266,148 @@ void Player::rediscover_node(const Map& map, const FCoords& f) {
+			throw wexception("duplicate seer entry for %s %u", m.descr().name().c_str(), m.serial());
 		}
 	}
+#endif
+	seers_.push_back(&m);
+}
 
-	Field& field = fields_[f.field - &first_map_field];
-	assert(fields_.get() <= &field);
-	assert(&field < fields_.get() + map.max_index());
-
-	if (field.vision == 0) {
-		field.vision = 1;
-	}
-	if (field.vision == 1) {
-		rediscover_node(map, f);
+void Player::remove_seer(const MapObject& m, const Area<FCoords>& a) {
+	for (auto it = seers_.begin(); it != seers_.end(); ++it) {
+		if (*it == &m) {
+			seers_.erase(it);
+			if (!egbase().objects().is_cleaning_up()) {
+				for (const PlayerNumber& p : team_player_) {
+					egbase().get_player(p)->update_vision(a, false);
+				}
+			}
+			return;
+		}
 	}
-	return ++field.vision;
+	NEVER_HERE();
 }
 
-/// If 'mode' = UnseeMode::kUnexplore, fields will be marked as unexplored. Else, player no longer
-/// sees what's currently going on. Returns the vision that this node had before it was hidden.
-Vision Player::unsee_node(MapIndex const i,
-                          Time const gametime,
-                          const SeeUnseeNode mode,
-                          bool const forward) {
-	Field& field = fields_[i];
-	if ((mode == SeeUnseeNode::kUnsee && field.vision <= 1) ||
-	    field.vision < 1) {  //  Already does not see this
-		return field.vision;
+// Checks if any of our buildings or bobs, or one of one of our team mates, is seeing this node
+bool Player::should_see(const FCoords& f) const {
+	for (const PlayerNumber& p : team_player_) {
+		const Player& player = *egbase().get_player(p);
+		if (player.revealed_fields_.count(f)) {
+			return true;
+		}
+		for (const MapObject* mo : player.seers_) {
+			if (mo->descr().type() >= MapObjectType::BUILDING) {
+				upcast(const Building, b, mo);
+				assert(b);
+				if (b->is_seeing() &&
+				    egbase().map().calc_distance(f, b->get_position()) <= b->descr().vision_range()) {
+					return true;
+				}
+			} else {
+				// currently only buildings and bobs can see fields
+				upcast(const Bob, b, mo);
+				assert(b);
+				if (egbase().map().calc_distance(f, b->get_position()) <= b->descr().vision_range()) {
+					return true;
+				}
+			}
+		}
 	}
+	return false;
+}
 
-	const Vision original_vision = field.vision;
-
-	//  If this is not already a forwarded call, we should inform allied players
-	//  as well of this change.
-	if (!team_player_uptodate_) {
-		update_team_players();
+void Player::update_vision(const FCoords& f, bool force_visible) {
+	if (!fields_ || egbase().objects().is_cleaning_up()) {
+		return;
 	}
-	if (!forward && !team_player_.empty()) {
-		for (uint8_t j = 0; j < team_player_.size(); ++j) {
-			team_player_[j]->unsee_node(i, gametime, mode, true);
+	Player::Field& field = fields_[egbase().map().get_index(f)];
+	if (force_visible || should_see(f)) {
+		if (field.seeing != SeeUnseeNode::kReveal) {
+			field.seeing = SeeUnseeNode::kReveal;
+			rediscover_node(egbase().map(), f);
```

Comment:
No `field.time_node_last_unseen = egbase().get_gametime();` here?

Reply:
That value needs to be updated only when changing the visibility state from `kReveal` to `kUnsee`. It is ignored for visible and unexplored fields.